### PR TITLE
Reimplemented preferences to store them in .blend

### PR DESCRIPTION
**Pull request by allista**
Also 
- Replaced the `API key` dialog with the `Wakatime Preferences` dialog
- Added `Download Wakatime` action to force re-download the wakatime client
- Separated various components into modules/classes in `wakatime_blender` main module.
  The `__init__.py` entry point only (un)registers things now.
- Updated the README

### CODE REVIEW

**Comment by allista** [comment]
@alanhamlett @williamchange would you be so kind as to glance at the changes?

@alanhamlett should the "Heartbeat Frequency" option be available to users?

**Comment by allista** [comment]
The [updated README](https://github.com/allista/WakatimeBlender/blob/v2/README.md) should help

File: __init__.py [hunk 1]
```diff
@@ -0,0 +1,104 @@
...
+    WakatimeProjectProperties,
+)
+from .wakatime_blender.wakatime_downloader import (
+    ForceWakatimeDownload,
+    WakatimeDownloader,
+)
+
+
+bl_info = {
+    "name": "Wakatime plugin for Blender",
```
Comment (alanhamlett) [inline __init__.py:18]: Can this be `"name": "WakaTime",`?

File: wakatime_blender/heartbeat_queue.py [hunk 2]
```diff
@@ -0,0 +1,178 @@
...
+            "--entity",
+            heartbeat.entity,
+            "--time",
+            f"{heartbeat.timestamp:f}",
+            "--plugin",
+            ua,
+        ]
+        props = WakatimeProjectProperties.instance()
+        if props and props.always_overwrite_name:
+            cmd.extend(["--project", heartbeat.project])
```
Comment (alanhamlett) [inline wakatime_blender/heartbeat_queue.py:116]: Usually we use `--alternate-project` to let wakatime-cli try detecting the project first. That keeps all the plugins consistent when they're working in the same project.

**Review by alanhamlett — CHANGES_REQUESTED** [review]
Can we follow [this example](https://github.com/wakatime/sublime-wakatime/blob/e5331d3086c6de2014b688553529aa1cf85d22b3/WakaTime.py#L619) and use the new Go wakatime-cli downloaded from GitHub releases instead of the old Python client?

When checking the current release version from the GitHub API, we use a header `Last-Modified` taken from `~/.wakatime-internal.cfg` and if GitHub returns a 304 then we use the last version cached in `~/.wakatime-internal.cfg`. That's because GitHub won't return the current version since it hasn't changed, and it won't count against the current IP's GitHub API public rate limit.

File: wakatime_blender/heartbeat_queue.py (hunk 2 above)
Comment (allista) [reply wakatime_blender/heartbeat_queue.py:116]: I've described this behaviour in the REDME here:

https://github.com/allista/WakatimeBlender/blob/v2/README.md#configuration

The thing is, usually there's no project per se with blender, just a file. No git either. So this option is added specifically for cases when wakatime cannot detect the name of the de-facto absent project.

File: __init__.py (hunk 1 above)
Comment (allista) [reply __init__.py]: Done

**Comment by allista** [comment]
Ok, will try this. Used the old one because didn't want to dive into os-arch detection, but dune it's already there...
But I'd recommend your team to make a separate python lib with this (and other common) logic, so that all python plugins (I see there are quite a few) are on the common ground.

**Comment by alanhamlett** [comment]
> I'd recommend your team to make a separate python lib with this (and other common) logic, so that all python plugins (I see there are quite a few) are on the common ground.

We can make a library but the code would still need to be bundled inline for most plugins, since plugins don't usually support installing dependencies.

**Comment by allista** [comment]
> We can make a library but the code would still need to be bundled inline for most plugins, since plugins don't usually support installing dependencies.

Hm... true, didn't think about that. Then maybe convert the legacy python wakatime client into a thin wrapper that determines the os/arch, then downloads and caches the real golang client?

In any case, for now I just copy the logic from sublime plugin as you suggested.

File: wakatime_blender/heartbeat_queue.py (hunk 2 above)
Comment (alanhamlett) [reply wakatime_blender/heartbeat_queue.py:116]: In that case when no project is detected, `--alternate-project` works the same way as `--project` but alternate is safer.

Comment (allista) [reply wakatime_blender/heartbeat_queue.py:116]: That's why the setting is called "always overwrite" and there's the `else:` clause below which does use `--alternate-project`. This gives a user more control over their statistics.

**Comment by allista** [comment]
> Can we follow [this example](https://github.com/wakatime/sublime-wakatime/blob/e5331d3086c6de2014b688553529aa1cf85d22b3/WakaTime.py#L619) and use the new Go wakatime-cli downloaded from GitHub releases instead of the old Python client?
> 
> When checking the current release version from the GitHub API, we use a header `Last-Modified` taken from `~/.wakatime-internal.cfg` and if GitHub returns a 304 then we use the last version cached in `~/.wakatime-internal.cfg`. That's because GitHub won't return the current version since it hasn't changed, and it won't count against the current IP's GitHub API public rate limit.

Tried to implement this in my code. To tell the truth, I don't like it: the amount of finely tuned custom logic used for the sole purpose of getting the _right_ client os+arch+version seems ridiculous compared to the logic of the plugin itself. IMHO, this should not be a part of any plugin, but, as I suggested earlier, rather be either the library, or a thin wrapper common to all python plugins.

So for now I'll merge the current version as is (with corrections already made) and leave this part for future versions.

**Pull request merged**